code/datums/components/attachment_handler.dm

```diff
@@ -0,0 +1,304 @@
+/datum/component/attachment_handler/proc/finish_handle_attachment(obj/item/attachment, list/input_attachment_data, mob/attacker)
+	var/list/attachment_data = input_attachment_data
+	if(!input_attachment_data)
+		attachment_data = get_attachment_data(attachment)
+
+	if(slots[attachment_data["slot"]])
+		var/obj/item/current_attachment = slots[attachment_data["slot"]]
+		finish_detach(current_attachment, get_attachment_data(current_attachment), attacker)
+
+	attachment.forceMove(parent)
+	slots[attachment_data["slot"]] = attachment
+
+	on_attach?.Invoke(attachment, attacker)
+	var/datum/callback/attachment_on_attach = attachment_data["on_attach"]
+	attachment_on_attach?.Invoke(parent, attacker)
+
+	update_overlay_icon_state(attachment, attachment_data["overlay_icon_state"])
+
+	update_parent_overlay()
+
+///This is the do_after and user checks for attaching.
+/datum/component/attachment_handler/proc/do_attach(obj/item/attachment, mob/attacher, list/attachment_data)
+	if(!ishuman(attacher))
+		return FALSE
+
+	var/mob/living/carbon/human/user = attacher
+
+	if(!can_attach(attachment, user, attachment_data))
+		return FALSE
+
+	var/obj/item/in_hand = user.get_inactive_held_item()
+	if(in_hand != parent)
+		to_chat(user, span_warning("You have to hold [parent] to do that!"))
+		return FALSE
+
+	var/idisplay = BUSY_ICON_GENERIC
+	var/attach_delay = attachment_data["attach_delay"]
+
+	var/skill_used = attachment_data["attach_skill"]
+	var/skill_upper_threshold = attachment_data["attach_skill_upper_threshold"]
+
+	if(skill_used)
+		if(user.skills.getRating(skill_used))
+			user.visible_message(span_notice("[user] begins attaching [attachment] to [parent]."),
+			span_notice("You begin attaching [attachment] to [parent]."), null, 4)
+			if(skill_upper_threshold && user.skills.getRating(skill_used) >= skill_upper_threshold) //See if the attacher is super skilled/panzerelite born to defeat never retreat etc
+				attach_delay *= 0.5
+		else //If the user has no training, attaching takes twice as long and they fumble about.
+			attach_delay *= 2
+			user.visible_message(span_notice("[user] begins fumbling about, trying to attach [attachment] to [parent]."),
+			span_notice("You begin fumbling about, trying to attach [attachment] to [parent]."), null, 4)
+			idisplay = BUSY_ICON_UNSKILLED
+
+	if(!do_after(user, attach_delay, TRUE, parent, idisplay))
+		return FALSE
+	
+
+
+	user.visible_message(span_notice("[user] attaches [attachment] to [parent]."),
+	span_notice("You attach [attachment] to [parent]."), null, 4)
+	playsound(user, attachment_data["attach_sound"], 15, 1, 4)
+	return TRUE
+
+///Checks the current slots of the parent and if there are attachments that can be removed in those slots. Basically it makes sure theres room for the attachment. 
+/datum/component/attachment_handler/proc/can_attach(obj/item/attachment, mob/living/carbon/human/user, list/attachment_data)
+	if(!slots.len)
```

Comment:
```suggestion
	if(!length(slots))
```

Reply:
it's not needed unless there is a chance that slots is null. not the case here